### PR TITLE
Allow Optimizers to perform global gradient clipping

### LINKED ISSUE #36001: Allow Optimizers to perform global gradient clipping
**System information**
- TensorFlow version: 2.1
- Are you willing to contribute it: Yes

**Describe the feature and the current behavior/state.**

Currently, passing `clipnorm` to a `tf.keras.optimizers.Optimizer` makes it clip the gradient for each weight tensor locally, or *independently of other weight gradients*:

https://github.com/tensorflow/tensorflow/blob/d8bcf5e4a24f9904e9a97b41da393e53149e999e/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L357-L358

I am proposing clipping the weight gradient *globally*, i.e.,

```python
    if hasattr(self, 'clipnorm'):
      grads, _ = clip_ops.clip_by_global_norm(grads, self.clipnorm)
```

This was already proposed in #29108 and #29114.

This change has at least two benefits:

1. Theoretical: the per-batch descent direction is preserved when gradients are clipped globally, but not when clipped locally.
2. Practical: [standalone Keras implements global gradient clipping](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/optimizers.py#L98-L100):
    ```python
        if hasattr(self, 'clipnorm') and self.clipnorm > 0:
            norm = K.sqrt(sum([K.sum(K.square(g)) for g in grads]))
            grads = [clip_norm(g, self.clipnorm, norm) for g in grads]
    ```
    so this change will unify `keras` and `tf.keras` behavior.

In fact, clipping by global norm is the behavior I *expected* when I first learned about the `clipnorm` parameter. The documentation does not clarify whether gradients are clipped locally or globally:

https://github.com/tensorflow/tensorflow/blob/d8bcf5e4a24f9904e9a97b41da393e53149e999e/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L241-L245

**Will this change the current api? How?**

The API will remain the same. Only the weight update behavior will change when the user passes `clipnorm` to an optimizer.

**Who will benefit with this feature?**

Users who use `clipnorm` with their optimizers.

**Any Other info.**

One of the first papers introducing gradient clipping by norm is [Pascanu, Mikolov, Bengio (2013)](http://proceedings.mlr.press/v28/pascanu13.html), where they propose
Screen Shot 2020-01-17 at 12 28 38 PM

In Section 1.1, θ is defined to be the collection of *all* weights in the model, not just a particular weight matrix or bias vector. This addresses [a question](https://github.com/tensorflow/tensorflow/pull/29114/files#r294862824
) In #29114 about whether the paper proposes global clipping or local clipping. Moreover, following Algorithm 1, Pascanu *et al* write,

> we only diverged from the original proposal in an attempt to provide a better theoretical justification (see section 2.3; we also move in a descent direction for the current mini-batch)

so they appear to be advocating altering the gradient in a way that preserves the descent direction (i.e., using global clipping).

Other recent sources also appear to suggest *global* gradient clipping (e.g., see equation (5) of [this ICLR 2020 paper](https://openreview.net/forum?id=BJgnXpVYwS)).

### PR DESCRIPTION
**Pull request by artemmavrin**
Fixes #36001

### CODE REVIEW

**Comment by gbaned** [comment]
@artemmavrin Can you please resolve conflicts? Thanks!

**Comment by artemmavrin** [comment]
@gbaned Conflicts removed!

**Comment by reedwm** [comment]
@fchollet @omalleyt12, can you review? I'm not sure how feasible it is to change this behavior at this point, but as #36001 mentions, this does make the behavior more consistent with standalone Keras.

**Comment by tanzhenyu** [comment]
I **think** we should fix it

**Comment by gbaned** [comment]
@tanzhenyu Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@tanzhenyu Any update on this PR? Please. Thanks!

**Comment by omalleyt12** [comment]
@artemmavrin Thanks for the PR!

Changing clipping to do clipping by global norm would be a backwards-incompatible change. However, we're in the process of making it much easier to provide custom gradient clipping and other gradient modifications, please check out the RFC here:

https://github.com/tensorflow/community/pull/234
